### PR TITLE
fix: prevent memory leak in Visualizer component

### PR DESCRIPTION
**Pull request by huyenltnguyen**
This PR:
- Is an attempt to reduce the CPU consumption by fixing a memory leak issue in the Visualizer component
- Related to #112

---
The issue can be reproduce by making the following changes to the code:
- Make the Main component render the Visualizer conditionally based on the `playing` state
- Add a console.log statement to [this line](https://github.com/freeCodeCamp/coderadio-client/blob/master/src/components/Visualizer.js#L88)

**Expected behavior:** When the user presses pause, the Visualizer component is unmounted, and there shouldn't be anything logged to the browser console.
**Actual behavior:** When the user presses pause, the Visualizer component is unmounted, and there are logs constantly printed in the browser console.

---
My fix for this is to move the recursion outside of the draw function, and implement `start` and `stop` functions that helps initiate and clean up animation loops instead.

References:
- Handle requestAnimationFrame loops: https://gist.github.com/jacob-beltran/aa114af1b6fd5de866aa365e3763a90b
- Delay component unmount: https://medium.com/@tomaszferens/delay-unmounting-of-the-component-in-react-8d6f6e73cdc

### CODE REVIEW

**Comment by huyenltnguyen** [comment]
I haven't been able to intensively test the CPU usage, but I do hope that it reduces some :crossed_fingers:

**Comment by ahmadabdolsaheb** [comment]
These are the errors I get after playing/pausing and playing again
Screen Shot 2021-01-11 at 1 52 30 PM

**Comment by huyenltnguyen** [comment]
Hmm. I'm guessing that the AudioContext didn't finish resetting at the time we pressed play again.

I'm actually thinking about keeping the Visualizer in the DOM when the music is paused or when the user switches tab, but stop the draw function instead, so that we won't have to reinitialize the audio context again. I think this approach will help keep the animation transition smooth.

**Comment by huyenltnguyen** [comment]
@ahmadabdolsaheb I found that apparently `ifvisible.js` has a memory leak issue (https://github.com/serkanyersen/ifvisible.js/issues/50), so I decided to use [react-page-visibility](https://www.npmjs.com/package/react-page-visibility) instead.

This PR is ready for another round of review :slightly_smiling_face:

File: src/components/Visualizer.js [hunk 1]
```diff
@@ -19,10 +26,31 @@ export default class Visualizer extends React.Component {
   // of the audio context stuff AFTER the audio has been triggered.
   // We can't see it until
   // then anyway so it makes no difference to desktop.
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.playing && !this.state.eq.context) {
-      this.initiateEQ();
+  componentDidUpdate(prevProps, prevState) {
+    if (prevProps.playing === this.props.playing && prevState.isTabVisible === this.state.isTabVisible) {
```
Comment (ahmadabdolsaheb) [inline src/components/Visualizer.js:30]: should we do (prevProps.playing !== this.props.playing && prevState.isTabVisible !== this.state.isTabVisible) so all of the conditions can stay in one block.

File: src/components/Visualizer.js [hunk 2]
```diff
@@ -1,7 +1,13 @@
 import React from 'react';
 import PropTypes from 'prop-types';
+import PageVisibility from 'react-page-visibility';
+
+const DELAY = 500;
+
+export default class Visualizer extends React.PureComponent {
+  rafId = null;
+  timerId = null;
```
Comment (ahmadabdolsaheb) [inline src/components/Visualizer.js:9]: Since they are changing, should we set the to states?

Comment (ahmadabdolsaheb) [reply src/components/Visualizer.js:9]: not sure what rafId stands for.

**Comment by ahmadabdolsaheb** [comment]
Thanks for the pr. Here are my nitpicks.

**Pull request merged**